### PR TITLE
[SPARK-32432][SQL] Added support for reading ORC/Parquet files with SymlinkTextInputFormat

### PR DESCRIPTION
**Pull request by moomindani**
### What changes were proposed in this pull request?

This pull-request is to add support for reading ORC/Parquet files with SymlinkTextInputFormat in Apache Spark.

### Why are the changes needed?

Hive style symlink (SymlinkTextInputFormat) is commonly used in different analytic engines including prestodb and prestosql.
Currently SymlinkTextInputFormat works with JSON/CSV files but does not work with ORC/Parquet files in Apache Spark (and Apache Hive).
On the other hand, prestodb and prestosql support SymlinkTextInputFormat with ORC/Parquet files.
This pull-request is to add support for reading ORC/Parquet files with SymlinkTextInputFormat in Apache Spark.

See details in the JIRA.  SPARK-32432

### Does this PR introduce _any_ user-facing change?

Yes.
Currently Spark returns exceptions if users try to use SymlinkTextInputFormat with ORC/Parquet files.
With this patch, Spark can handle symlink which indicates locations of ORC/Parquet files.

### How was this patch tested?

I added a new test suite `SymlinkSuite` and confirmed it passed.

```
$ ./build/sbt "project hive" "test-only org.apache.spark.sql.hive.SymlinkSuite"
```

### CODE REVIEW

**Comment by moomindani** [comment]
retest this please

**Comment by moomindani** [comment]
retest this please

**Comment by moomindani** [comment]
Could you please take a look and review this PR?

**Comment by manuzhang** [comment]
cc @cloud-fan @gengliangwang @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SymlinkTextInputFormatUtil.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  def isSymlinkTextFormat(inputFormat: String): Boolean = {
+    inputFormat.equals("org.apache.hadoop.hive.ql.io.SymlinkTextInputFormat")
+  }
+
+  // Mostly copied from SymlinkTextInputFormat#getTargetPathsFromSymlinksDirs of Hive 3.1
+  def getTargetPathsFromSymlink(
+      fileSystem: FileSystem,
+      symlinkDir: Path): Seq[Path] = {
+
+    val symlinkIterator = fileSystem.listFiles(symlinkDir, true)
```
Comment (CHENXCHEN) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SymlinkTextInputFormatUtil.scala:39]: Hi, you should ignore the name start with "\_" or ".", just like this [hive implement](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/io/SymlinkTextInputFormat.java#L191), cause hadoop will generate temp file start with "\_" or ".", we should ignore these files.

**Review by CHENXCHEN — CHANGES_REQUESTED** [review]
Temporary files generated by hadoop will be read. @moomindani

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SymlinkTextInputFormatUtil.scala (hunk 1 above)
Comment (moomindani) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SymlinkTextInputFormatUtil.scala:39]: Thanks for the comments, I will reflect your comments on this.

**Comment by moomindani** [comment]
Let me keep this PFR open.

**Comment by maropu** [comment]
ok to test

**Comment by gatorsmile** [comment]
cc @c21 Is it supported in FB?

**Comment by c21** [comment]
@gatorsmile - we indeed support it, but the number of query everyday is very minimal, after checking the data.

**Comment by gatorsmile** [comment]
@c21 Can you help review the PR or open source your internal fix?

**Comment by c21** [comment]
@gatorsmile - sure I will take a look, thanks.

**Comment by jles01** [comment]
Hi,

Is there any updates on this ?

**Review by sadikovi — CHANGES_REQUESTED** [review]
There are not that many details on the JIRA ticket and the provided links don't work - would appreciate it if you could update the description. I will do another pass once the inline comments are addressed

**Pull request closed without merging**